### PR TITLE
Smarter thermal controls for tici

### LINKED ISSUE #21917: Improved C3 fan policy

### PR DESCRIPTION
**Pull request by robbederks**

### CODE REVIEW

File: selfdrive/thermald/thermald.py [hunk 1]
```diff
@@ -216,9 +240,12 @@ def thermald_thread():
       if handle_fan is None and pandaState.pandaState.pandaType != log.PandaState.PandaType.unknown:
         is_uno = pandaState.pandaState.pandaType == log.PandaState.PandaType.uno
 
-        if (not EON) or is_uno:
+        if (not EON and not TICI) or is_uno:
           cloudlog.info("Setting up UNO fan handler")
           handle_fan = handle_fan_uno
+        elif TICI:
+          cloudlog.info("Setting up TICI fan handler")
+          handle_fan = handle_fan_tici
```
Comment (adeebshihadeh) [inline selfdrive/thermald/thermald.py:248]:
```suggestion
        if TICI:
          cloudlog.info("Setting up TICI fan handler")
          handle_fan = handle_fan_tici
        elif is_uno:
          cloudlog.info("Setting up UNO fan handler")
          handle_fan = handle_fan_uno
```

File: selfdrive/thermald/thermald.py [hunk 2]
```diff
@@ -121,6 +121,30 @@ def handle_fan_uno(max_cpu_temp, bat_temp, fan_speed, ignition):
...
+def handle_fan_tici(max_cpu_temp, bat_temp, fan_speed, ignition):
+  global prev_time, error_integral, prev_cpu_temp_error
+
+  CPU_TEMP_SETPOINT = 80 if ignition else 60
+  FAN_PWR_RANGE = (0, 80 if ignition else 30)
+  KI = 1e-5
+
+  if prev_time is None:
+    prev_time = time.monotonic()
+  dt = max(time.monotonic() - prev_time, 0.01)
```
Comment (adeebshihadeh) [inline selfdrive/thermald/thermald.py:135]: does this work? looks like prev_time is only set once

Comment (robbederks) [reply selfdrive/thermald/thermald.py:135]: lol, that's why it gets unstable after long drives, good catch!

File: selfdrive/thermald/thermald.py [hunk 3]
```diff
@@ -104,16 +103,12 @@ def handle_fan_eon(max_cpu_temp, bat_temp, fan_speed, ignition):
     # update speed if using the low thresholds results in fan speed decrement
     fan_speed = new_speed_l
 
-  if bat_temp < _BAT_TEMP_THRESHOLD:
```
Comment (adeebshihadeh) [inline selfdrive/thermald/thermald.py:107]: are you sure we can remove this? we still support the EON gold

File: selfdrive/controls/lib/pid.py [hunk 4]
```diff
@@ -1,3 +1,4 @@
+from numbers import Number
 import numpy as np
 from common.numpy_fast import clip, interp
```
Comment (adeebshihadeh) [inline selfdrive/controls/lib/pid.py:3]:
```suggestion
import numpy as np
from numbers import Number

from common.numpy_fast import clip, interp
```

File: selfdrive/controls/lib/pid.py [hunk 5]
```diff
@@ -29,10 +31,14 @@ def __init__(self, k_p, k_i, k_f=1., pos_limit=None, neg_limit=None, rate=100, s
 
   @property
   def k_p(self):
+    if isinstance(self._k_p, Number):
+      return self._k_p
```
Comment (adeebshihadeh) [inline selfdrive/controls/lib/pid.py:35]: we shouldn't handle two types here. why not just pass in a list?

Comment (robbederks) [reply selfdrive/controls/lib/pid.py:35]: Just a list doesn't work, needs to be a list of lists containing x,y pairs for interpolation based on speed (which doesn't make much sense outside of a vehicle controls context). Had it this way first, but looks pretty ugly

File: selfdrive/thermald/thermald.py (hunk 3 above)
Comment (robbederks) [reply selfdrive/thermald/thermald.py:107]: this is fine, worst case the EON fans might spin a little faster

**Pull request merged**